src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java

```diff
@@ -65,47 +65,76 @@ public OpticalDuplicateFinder(final String readNameRegex, final int opticalDupli
+
+        final PhysicalLocation actualKeeper = keeperOrNull(list, keeper);
 
-        Collections.sort(list, new Comparator<PhysicalLocation>() {
-            public int compare(final PhysicalLocation lhs, final PhysicalLocation rhs) {
-                int retval = lhs.getReadGroup() - rhs.getReadGroup();
-                if (retval == 0) retval = lhs.getTile() - rhs.getTile();
-                if (retval == 0) retval = lhs.getX() - rhs.getX();
-                if (retval == 0) retval = lhs.getY() - rhs.getY();
-                return retval;
+        // First go through and compare all the reads to the keeper
+        if (actualKeeper != null) {
+            for (int i=0; i<length; ++i) {
+                final PhysicalLocation other = list.get(i);
+                opticalDuplicateFlags[i] = closeEnough(actualKeeper, other, distance);
             }
-        });
+        }
 
-        outer:
-        for (int i = 0; i < length; ++i) {
+        // Now go through and do each pairwise comparison not involving the actualKeeper
+        for (int i=0; i<length; ++i) {
             final PhysicalLocation lhs = list.get(i);
-            if (lhs.getTile() < 0) continue;
+            if (lhs == actualKeeper) continue; // no comparisons to actualKeeper since those are all handled above
 
-            for (int j = i + 1; j < length; ++j) {
+            for (int j =i+1; j<length; ++j) {
                 final PhysicalLocation rhs = list.get(j);
+                if (rhs == actualKeeper) continue; // no comparisons to actualKeeper since those are all handled above
 
-                if (opticalDuplicateFlags[j]) continue;
-                if (lhs.getReadGroup() != rhs.getReadGroup()) continue outer;
-                if (lhs.getTile() != rhs.getTile()) continue outer;
-                if (rhs.getX() > lhs.getX() + this.opticalDuplicatePixelDistance) continue outer;
-
-                if (Math.abs(lhs.getY() - rhs.getY()) <= this.opticalDuplicatePixelDistance) {
-                    opticalDuplicateFlags[j] = true;
+                if (closeEnough(lhs, rhs, distance)) {
+                    // At this point we want to mark either lhs or rhs as duplicate. Generally we want to mark
+                    // rhs, unless rhs is already marked.  The only way rhs can already be marked is if it's a
+                    // duplicate of the 'keeper' (i.e. and a keeper was provided), so in that case it's safe
+                    // to mark the lhs instead, and so we do.
+                    final int index = opticalDuplicateFlags[j] ? i : j;
+                    opticalDuplicateFlags[index] = true;
                 }
             }
         }
+        
         return opticalDuplicateFlags;
     }
+
+    /** Returns the keeper if it is contained within the list and has location information, otherwise null. */
+    private PhysicalLocation keeperOrNull(final List<? extends PhysicalLocation> list, final PhysicalLocation keeper) {
```

Comment:
I agree that this could be static, but I see no advantage to making it static - it's only referenced from a non-static context.